python/lsst/pipe/base/graph/_implDetails.py

```diff
@@ -181,3 +233,119 @@ def __contains__(self, key: _T) -> bool:
+                                     dataId=node.quantum.dataId, initInputs=node.quantum.initInputs,
+                                     inputs=helper.inputs, outputs=helper.outputs)
+                # If the inputs or outputs were adjusted to something different
+                # than what was supplied by the graph builder, dissassociate
+                # node from those refs, and if they are output refs, prune them
+                # from downstream tasks. This means that based on new inputs
+                # the task wants to produce fewer outputs, or consume fewer
+                # inputs.
+                for condition, existingMapping, newMapping, remover in (
+                        (helper.inputs_adjusted, node.quantum.inputs, helper.inputs,
+                         datasetRefDict.removeConsumer),
+                        (helper.outputs_adjusted, node.quantum.outputs, helper.outputs,
+                         datasetRefDict.removeProducer)):
+                    if condition:
+                        notNeeded = set()
+                        for key in existingMapping:
+                            if key not in newMapping:
+                                compositeRefs = (r if not r.isComponent() else r.makeCompositeRef()
+                                                 for r in existingMapping[key])
+                                notNeeded |= set(compositeRefs)
+                                continue
+                            notNeeded |= set(existingMapping[key]) - set(newMapping[key])
+                        if notNeeded:
+                            for ref in notNeeded:
+                                if ref.isComponent():
+                                    ref = ref.makeCompositeRef()
+                                remover(ref, node)
+                            if remover is datasetRefDict.removeProducer:
+                                _pruner(datasetRefDict, notNeeded, alreadyPruned=alreadyPruned)
+                object.__setattr__(node, 'quantum', newQuantum)
+
+            except NoWorkFound:
+                # This will throw if the length is less than the minimum number
+                for tmpRef in chain(chain.from_iterable(node.quantum.inputs.values()),
+                                    node.quantum.initInputs.values()):
+                    if tmpRef.isComponent():
+                        tmpRef = tmpRef.makeCompositeRef()
+                    datasetRefDict.removeConsumer(tmpRef, node)
+                alreadyPruned.add(node)
+                # prune all outputs produced by this node
+                # mark that none of these will be produced
+                forwardPrunes = set()
+                for forwardRef in chain.from_iterable(node.quantum.outputs.values()):
+                    datasetRefDict.removeProducer(forwardRef, node)
+                    forwardPrunes.add(forwardRef)
+                _pruner(datasetRefDict, forwardPrunes, alreadyPruned=alreadyPruned)
```

Comment:
I'm a little wary of recursing (as well as the logic earlier) inside the `except` block, as it seems like a way to get ugly "while handling an exception, another exception was raised" tracebacks when something goes wrong.  Could we put a `return` at the end of `try` and just de-indent most of the `except` block?